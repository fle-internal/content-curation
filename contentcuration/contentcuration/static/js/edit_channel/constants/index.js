const ContentKinds = require('./ContentKinds');
const FormatPresets = require('./FormatPresets');
const Languages = require('./Languages');
const Licenses = require('./Licenses');
const MasteryModels = require('./MasteryModels');
const Roles = require('./Roles');

module.exports = {
  ContentKinds,
  FormatPresets,
  Languages,
  Licenses,
  MasteryModels,
  Roles,
};
