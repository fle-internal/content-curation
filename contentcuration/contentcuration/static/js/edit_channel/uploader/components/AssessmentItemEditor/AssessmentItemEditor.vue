<template>
  <div>
    <VLayout>
      <VFlex xs7 lg5>
        <VSelect
          :key="kindSelectKey"
          :items="kindSelectItems"
          :value="kind"
          :label="$tr('questionTypeLabel')"
          data-test="kindSelect"
          @input="onKindUpdate"
        />
      </VFlex>
    </VLayout>

    <VLayout>
      <VFlex>
        <ErrorList
          :errors="questionErrors"
          data-test="questionErrors"
        />

        <div class="grey--text text--darken-2 mb-1">
          {{ $tr('questionLabel') }}
        </div>

        <transition name="fade">
          <keep-alive include="MarkdownEditor">
            <MarkdownEditor
              v-if="isQuestionOpen"
              :markdown="question"
              @update="onQuestionUpdate"
            />

            <div
              v-else
              class="pl-2 pr-2 pt-3 pb-3 question-text"
              data-test="questionText"
              @click="openQuestion"
            >
              <VIcon color="grey darken-1">
                edit
              </VIcon>
              <span>{{ question }}</span>
            </div>
          </keep-alive>
        </transition>
      </VFlex>
    </VLayout>

    <VLayout mt-4>
      <VFlex>
        <ErrorList
          :errors="answersErrors"
          data-test="answersErrors"
        />

        <AnswersEditor
          :questionKind="kind"
          :answers="answers"
          :openAnswerIdx="openAnswerIdx"
          @update="onAnswersUpdate"
          @open="openAnswer"
          @close="closeAnswer"
        />

        <HintsEditor
          class="mt-4"
          :hints="hints"
          :openHintIdx="openHintIdx"
          @update="onHintsUpdate"
          @open="openHint"
          @close="closeHint"
        />
      </VFlex>
    </VLayout>
  </div>
</template>

<script>

  import { AssessmentItemTypes, AssessmentItemTypeLabels } from '../../constants';
  import { getAssessmentItemErrorMessage, updateAnswersToQuestionType } from '../../utils';

  import AnswersEditor from '../AnswersEditor/AnswersEditor.vue';
  import ErrorList from '../ErrorList/ErrorList.vue';
  import HintsEditor from '../HintsEditor/HintsEditor.vue';
  import MarkdownEditor from '../MarkdownEditor/MarkdownEditor.vue';

  export default {
    name: 'AssessmentItemEditor',
    $trs: {
      questionTypeLabel: 'Question type',
      questionLabel: 'Question',
      dialogTitle: 'Changing question type',
      dialogSubmitBtnLabel: 'Change',
      dialogMessageChangeToSingleSelection:
        'Switching to single selection will set only one answer as correct. Continue?',
      dialogMessageChangeToTrueFalse:
        'Switching to true or false will remove any current answers. Continue?',
      dialogMessageChangeToInput:
        'Switching to numeric input will set all answers as correct and remove all non-numeric answers. Continue?',
    },
    components: {
      AnswersEditor,
      ErrorList,
      HintsEditor,
      MarkdownEditor,
    },
    model: {
      prop: 'item',
      event: 'update',
    },
    props: {
      /**
       * {
       *   // assessment item data as retrieved from API
       *   data: {
       *      question
       *      type
       *      order
       *      answers
       *      hints
       *      ...
       *   },
       *   // client validation data for the assessment item
       *   validation: {
       *      questionErrors
       *      answerErrors
       *   }
       * }
       */
      item: {
        type: Object,
      },
      /**
       * Inject a function that opens a dialog that should
       * be confirmed before certain actions can be performed.
       * If not provided, no confirmation will be required.
       * Expected interface:
       *   openDialog({
       *     title: String,
       *     message: String,
       *     cancelLabel: String,
       *     submitLabel: String,
       *     onCancel: Function,
       *     onSubmit: Function,
       *   })
       * })
       */
      openDialog: {
        type: Function,
      },
    },
    data() {
      return {
        isQuestionOpen: false,
        openHintIdx: null,
        openAnswerIdx: null,
        kindSelectKey: 0,
      };
    },
    computed: {
      itemData() {
        if (!this.item) {
          return null;
        }

        return this.item.data;
      },
      itemValidation() {
        if (!this.item) {
          return null;
        }

        return this.item.validation;
      },
      question() {
        if (!this.itemData || !this.itemData.question) {
          return '';
        }

        return this.itemData.question;
      },
      kind() {
        if (!this.itemData || !this.itemData.type) {
          return AssessmentItemTypes.SINGLE_SELECTION;
        }

        return this.itemData.type;
      },
      kindSelectItems() {
        return [
          {
            value: AssessmentItemTypes.SINGLE_SELECTION,
            text: AssessmentItemTypeLabels[AssessmentItemTypes.SINGLE_SELECTION],
          },
          {
            value: AssessmentItemTypes.MULTIPLE_SELECTION,
            text: AssessmentItemTypeLabels[AssessmentItemTypes.MULTIPLE_SELECTION],
          },
          {
            value: AssessmentItemTypes.INPUT_QUESTION,
            text: AssessmentItemTypeLabels[AssessmentItemTypes.INPUT_QUESTION],
          },
          {
            value: AssessmentItemTypes.TRUE_FALSE,
            text: AssessmentItemTypeLabels[AssessmentItemTypes.TRUE_FALSE],
          },
        ];
      },
      answers() {
        if (!this.itemData || !this.itemData.answers) {
          return [];
        }

        return this.itemData.answers;
      },
      hints() {
        if (!this.itemData || !this.itemData.hints) {
          return [];
        }

        return this.itemData.hints;
      },
      questionErrors() {
        if (!this.itemValidation || !this.itemValidation.questionErrors) {
          return [];
        }

        return this.itemValidation.questionErrors.map(error =>
          getAssessmentItemErrorMessage(error, this.kind)
        );
      },
      answersErrors() {
        if (!this.itemValidation || !this.itemValidation.answersErrors) {
          return [];
        }

        return this.itemValidation.answersErrors.map(error =>
          getAssessmentItemErrorMessage(error, this.kind)
        );
      },
      hasMoreCorrectAnswers() {
        const correctAnswers = this.answers.filter(answer => answer.correct === true);

        return correctAnswers.length > 1;
      },
    },
    mounted() {
      if (!this.question) {
        this.openQuestion();
      }
    },
    methods: {
      updateItem(data) {
        let newItem = {
          ...this.item,
          data: {
            ...this.item.data,
            ...data,
          },
        };

        this.$emit('update', newItem);
      },
      changeKind(newKind) {
        const newAnswers = updateAnswersToQuestionType(newKind, this.answers);

        this.closeAnswer();
        this.updateItem({ type: newKind, answers: newAnswers });
      },
      // question type VSelect needs to be rerended when confirmation dialog
      // cancelled to display a correct, previous, value that has changed
      // in the select but has not been changed in data storage actually
      // because of cancel action
      rerenderKindSelect() {
        this.kindSelectKey += 1;
      },
      onQuestionUpdate(newQuestion) {
        this.updateItem({ question: newQuestion });
      },
      onKindUpdate(newKind) {
        if (this.kind === newKind) {
          return;
        }

        switch (newKind) {
          case AssessmentItemTypes.SINGLE_SELECTION:
            if (typeof this.openDialog === 'function' && this.hasMoreCorrectAnswers) {
              this.openDialog({
                title: this.$tr('dialogTitle'),
                message: this.$tr('dialogMessageChangeToSingleSelection'),
                submitLabel: this.$tr('dialogSubmitBtnLabel'),
                onSubmit: () => this.changeKind(newKind),
                onCancel: this.rerenderKindSelect,
              });
            } else {
              this.changeKind(newKind);
            }

            break;

          case AssessmentItemTypes.TRUE_FALSE:
            if (typeof this.openDialog === 'function' && this.answers.length > 0) {
              this.openDialog({
                title: this.$tr('dialogTitle'),
                message: this.$tr('dialogMessageChangeToTrueFalse'),
                submitLabel: this.$tr('dialogSubmitBtnLabel'),
                onSubmit: () => this.changeKind(newKind),
                onCancel: this.rerenderKindSelect,
              });
            } else {
              this.changeKind(newKind);
            }

            break;

          case AssessmentItemTypes.INPUT_QUESTION:
            if (typeof this.openDialog === 'function' && this.answers.length > 0) {
              this.openDialog({
                title: this.$tr('dialogTitle'),
                message: this.$tr('dialogMessageChangeToInput'),
                submitLabel: this.$tr('dialogSubmitBtnLabel'),
                onSubmit: () => this.changeKind(newKind),
                onCancel: this.rerenderKindSelect,
              });
            } else {
              this.changeKind(newKind);
            }

            break;

          default:
            this.changeKind(newKind);
            break;
        }
      },
      onAnswersUpdate(newAnswers) {
        this.updateItem({ answers: newAnswers });
      },
      onHintsUpdate(newHints) {
        this.updateItem({ hints: newHints });
      },
      openQuestion() {
        this.isQuestionOpen = true;

        this.closeAnswer();
        this.closeHint();
      },
      closeQuestion() {
        this.isQuestionOpen = false;
      },
      openHint(hintIdx) {
        this.openHintIdx = hintIdx;

        this.closeQuestion();
        this.closeAnswer();
      },
      closeHint() {
        this.openHintIdx = null;
      },
      openAnswer(answerIdx) {
        this.openAnswerIdx = answerIdx;

        this.closeQuestion();
        this.closeHint();
      },
      closeAnswer() {
        this.openAnswerIdx = null;
      },
    },
  };

</script>

<style lang="less" scoped>

  @import '../../../../../less/global-variables.less';

  .question-text {
    transition: 0.7s;

    &:hover {
      cursor: pointer;
      background-color: @gray-light;
    }
  }

</style>
