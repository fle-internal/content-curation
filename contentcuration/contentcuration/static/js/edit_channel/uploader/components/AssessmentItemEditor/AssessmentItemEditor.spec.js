import { shallowMount, mount } from '@vue/test-utils';

import { AssessmentItemTypes, AssessmentItemValidationErrors } from '../../constants';
import AssessmentItemEditor from './AssessmentItemEditor';

jest.mock('../MarkdownEditor/MarkdownEditor.vue');

const ITEM = {
  data: {
    question: 'Exercise 2 - Question 2',
    type: AssessmentItemTypes.SINGLE_SELECTION,
    answers: [
      { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
      { answer: 'Peanut butter', correct: false, order: 2 },
    ],
    hints: [{ hint: "It's not healthy", order: 1 }, { hint: 'Tasty!', order: 2 }],
  },
  validation: {},
};

const openQuestion = wrapper => {
  wrapper.find('[data-test="questionText"]').trigger('click');
};

const updateQuestion = (wrapper, newQuestionText) => {
  // only one editor is rendered at a time => "wrapper.find"
  wrapper.find({ name: 'MarkdownEditor' }).vm.$emit('update', newQuestionText);
};

const selectKind = (wrapper, kind) => {
  const input = wrapper.find('[data-test="kindSelect"]');
  input.element.value = kind;

  input.trigger('input');
};

describe('AssessmentItemEditor', () => {
  let wrapper;

  it('smoke test', () => {
    const wrapper = shallowMount(AssessmentItemEditor);

    expect(wrapper.isVueInstance()).toBe(true);
  });

  it('renders', () => {
    wrapper = mount(AssessmentItemEditor, {
      propsData: {
        item: ITEM,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('on question text update', () => {
    beforeEach(() => {
      wrapper = mount(AssessmentItemEditor, {
        propsData: {
          item: ITEM,
        },
      });

      openQuestion(wrapper);
      updateQuestion(wrapper, 'My new question');
    });

    it('emits update event with item containing updated question text', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual({
        ...ITEM,
        data: {
          ...ITEM.data,
          question: 'My new question',
        },
      });
    });
  });

  describe('on item type update', () => {
    describe('when changing to single selection', () => {
      beforeEach(() => {
        const item = JSON.parse(JSON.stringify(ITEM));
        item.data = {
          ...item.data,
          type: AssessmentItemTypes.MULTIPLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: true, order: 2 },
          ],
        };

        wrapper = mount(AssessmentItemEditor, {
          propsData: {
            item,
          },
        });

        selectKind(wrapper, AssessmentItemTypes.SINGLE_SELECTION);
      });

      it('emits update event with item containing updated answers and type', () => {
        expect(wrapper.emitted().update).toBeTruthy();
        expect(wrapper.emitted().update.length).toBe(1);
        expect(wrapper.emitted().update[0][0]).toEqual({
          ...ITEM,
          data: {
            ...ITEM.data,
            answers: [
              { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
              { answer: 'Peanut butter', correct: false, order: 2 },
            ],
            type: AssessmentItemTypes.SINGLE_SELECTION,
          },
          validation: {},
        });
      });
    });

    describe('when changing to multiple selection', () => {
      beforeEach(() => {
        const item = JSON.parse(JSON.stringify(ITEM));
        item.data = {
          ...item.data,
          type: AssessmentItemTypes.SINGLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        };

        wrapper = mount(AssessmentItemEditor, {
          propsData: {
            item,
          },
        });

        selectKind(wrapper, AssessmentItemTypes.MULTIPLE_SELECTION);
      });

      it('emits update event with item containing same answers and type', () => {
        expect(wrapper.emitted().update).toBeTruthy();
        expect(wrapper.emitted().update.length).toBe(1);
        expect(wrapper.emitted().update[0][0]).toEqual({
          ...ITEM,
          data: {
            ...ITEM.data,
            answers: [
              { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
              { answer: 'Peanut butter', correct: false, order: 2 },
            ],
            type: AssessmentItemTypes.MULTIPLE_SELECTION,
          },
        });
      });
    });

    describe('when changing to true/false', () => {
      beforeEach(() => {
        const item = JSON.parse(JSON.stringify(ITEM));
        item.data = {
          ...item.data,
          type: AssessmentItemTypes.SINGLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        };

        wrapper = mount(AssessmentItemEditor, {
          propsData: {
            item,
          },
        });

        selectKind(wrapper, AssessmentItemTypes.TRUE_FALSE);
      });

      it('emits update event with item containing updated answers and type', () => {
        expect(wrapper.emitted().update).toBeTruthy();
        expect(wrapper.emitted().update.length).toBe(1);
        expect(wrapper.emitted().update[0][0]).toEqual({
          ...ITEM,
          data: {
            ...ITEM.data,
            answers: [
              { answer: 'True', order: 1, correct: true },
              { answer: 'False', order: 2, correct: false },
            ],
            type: AssessmentItemTypes.TRUE_FALSE,
          },
        });
      });
    });

    describe('when changing to input question', () => {
      beforeEach(() => {
        const item = JSON.parse(JSON.stringify(ITEM));
        item.data = {
          ...item.data,
          type: AssessmentItemTypes.SINGLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        };

        wrapper = mount(AssessmentItemEditor, {
          propsData: {
            item,
          },
        });

        selectKind(wrapper, AssessmentItemTypes.INPUT_QUESTION);
      });

      it('emits update event with item containing updated answers and type', () => {
        expect(wrapper.emitted().update).toBeTruthy();
        expect(wrapper.emitted().update.length).toBe(1);
        expect(wrapper.emitted().update[0][0]).toEqual({
          ...ITEM,
          data: {
            ...ITEM.data,
            answers: [
              { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
              { answer: 'Peanut butter', correct: true, order: 2 },
            ],
            type: AssessmentItemTypes.INPUT_QUESTION,
          },
        });
      });
    });
  });

  describe('on answers update', () => {
    beforeEach(() => {
      let item = JSON.parse(JSON.stringify(ITEM));
      item.data = {
        ...item.data,
        type: AssessmentItemTypes.SINGLE_SELECTION,
        answers: [
          { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
          { answer: 'Peanut butter', correct: false, order: 2 },
        ],
      };

      wrapper = mount(AssessmentItemEditor, {
        propsData: {
          item,
        },
      });

      const newAnswers = [
        { answer: 'Mayonnaise (I mean you can, but...)', correct: false, order: 1 },
        { answer: 'Peanut butter', correct: false, order: 2 },
      ];

      wrapper.find({ name: 'AnswersEditor' }).vm.$emit('update', newAnswers);
    });

    it('emits update event with an item containing updated answers', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual({
        ...ITEM,
        data: {
          ...ITEM.data,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: false, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        },
      });
    });
  });

  describe('on hints update', () => {
    beforeEach(() => {
      let item = JSON.parse(JSON.stringify(ITEM));
      item.data = {
        ...item.data,
        hints: [{ hint: 'Hint 1', order: 1 }],
      };

      wrapper = mount(AssessmentItemEditor, {
        propsData: {
          item,
        },
      });

      const newHints = [{ hint: 'Hint 1', order: 1 }, { hint: 'Hint 2', order: 2 }];

      wrapper.find({ name: 'HintsEditor' }).vm.$emit('update', newHints);
    });

    it('emits update event with item containing updated hints', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual({
        ...ITEM,
        data: {
          ...ITEM.data,
          hints: [{ hint: 'Hint 1', order: 1 }, { hint: 'Hint 2', order: 2 }],
        },
      });
    });
  });

  describe('for an invalid item', () => {
    beforeEach(() => {
      const item = JSON.parse(JSON.stringify(ITEM));
      (item.validation = {
        questionErrors: [AssessmentItemValidationErrors.BLANK_QUESTION],
        answersErrors: [AssessmentItemValidationErrors.INVALID_NUMBER_OF_CORRECT_ANSWERS],
      }),
        (wrapper = mount(AssessmentItemEditor, {
          propsData: {
            item,
          },
        }));
    });

    it('renders all errors messages', () => {
      expect(wrapper.find('[data-test=questionErrors]')).toMatchSnapshot();
      expect(wrapper.find('[data-test=answersErrors]')).toMatchSnapshot();
    });
  });
});
