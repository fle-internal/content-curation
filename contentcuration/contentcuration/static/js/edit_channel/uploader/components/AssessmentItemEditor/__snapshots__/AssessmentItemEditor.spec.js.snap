// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssessmentItemEditor for an invalid item renders all errors messages 1`] = `
<div role="list" class="v-list theme--light" data-test="questionErrors">
  <div role="listitem">
    <div class="v-list__tile theme--light"><span class="red--text font-weight-medium">Question cannot be blank</span></div>
  </div>
</div>
`;

exports[`AssessmentItemEditor for an invalid item renders all errors messages 2`] = `
<div role="list" class="v-list theme--light" data-test="answersErrors">
  <div role="listitem">
    <div class="v-list__tile theme--light"><span class="red--text font-weight-medium">Choose a correct answer</span></div>
  </div>
</div>
`;

exports[`AssessmentItemEditor renders 1`] = `
<div>
  <div class="layout">
    <div class="flex xs7 lg5">
      <div role="combobox" class="v-input v-text-field v-select v-input--is-label-active v-input--is-dirty theme--light">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-select__slot"><label aria-hidden="true" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Question type</label>
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">Single selection</div><input aria-label="Question type" data-test="kindSelect" readonly="readonly" type="text" autocomplete="on" aria-readonly="false">
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon material-icons theme--light">arrow_drop_down</i></div>
              </div>
            </div>
            <div class="v-menu">
              <div class="v-menu__content theme--light " style="max-height: 300px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;">
                <div class="v-select-list v-card theme--light">
                  <div role="list" class="v-list theme--light">
                    <div role="listitem" class="primary--text"><a class="v-list__tile v-list__tile--link v-list__tile--active theme--light primary--text">
                        <div class="v-list__tile__content">
                          <div class="v-list__tile__title">Single selection</div>
                        </div>
                      </a></div>
                    <div role="listitem"><a class="v-list__tile v-list__tile--link theme--light">
                        <div class="v-list__tile__content">
                          <div class="v-list__tile__title">Multiple selection</div>
                        </div>
                      </a></div>
                    <div role="listitem"><a class="v-list__tile v-list__tile--link theme--light">
                        <div class="v-list__tile__content">
                          <div class="v-list__tile__title">Input question</div>
                        </div>
                      </a></div>
                    <div role="listitem"><a class="v-list__tile v-list__tile--link theme--light">
                        <div class="v-list__tile__content">
                          <div class="v-list__tile__title">True/False</div>
                        </div>
                      </a></div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <div class="v-messages__wrapper"></div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="layout">
    <div class="flex">
      <!---->
      <div class="grey--text text--darken-2 mb-1">
        Question
      </div>
      <div data-test="questionText" class="pl-2 pr-2 pt-3 pb-3 question-text" name="fade"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">edit</i> <span>Exercise 2 - Question 2</span></div>
    </div>
  </div>
  <div class="layout mt-4">
    <div class="flex">
      <!---->
      <div>
        <div class="grey--text text--darken-1 mb-3">
          Answers
        </div>
        <!---->
        <div class="card-border-light">
          <div class="v-card v-card--flat v-sheet theme--light answer editable closed correct" data-test="answer">
            <div class="indicator correct"></div>
            <div class="v-card__text">
              <div class="layout align-top">
                <div class="flex">
                  <div class="v-input v-input--selection-controls v-input--radio-group v-input--radio-group--column theme--light">
                    <div class="v-input__control">
                      <div class="v-input__slot" style="height: auto;">
                        <div role="radiogroup" class="v-input--radio-group__input">
                          <div class="v-radio theme--light accent--text">
                            <div class="v-input--selection-controls__input"><input aria-label="Mayonnaise (I mean you can, but...)" aria-checked="true" role="radio" type="radio" name="v-radio-32" value="0" data-test="answerRadio">
                              <div class="v-input--selection-controls__ripple accent--text"></div><i aria-hidden="true" class="v-icon material-icons theme--light accent--text">radio_button_checked</i>
                            </div><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: relative;">Mayonnaise (I mean you can, but...)</label>
                          </div>
                        </div>
                      </div>
                      <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                    </div>
                  </div>
                  <!---->
                  <!---->
                </div>
                <div class="flex">
                  <!---->
                </div>
                <div class="spacer"></div>
                <div class="flex">
                  <div class="layout justify-end toolbar" data-test="toolbar">
                    <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move up</span></div><button disabled="disabled" type="button" class="v-btn v-btn--disabled v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_UP">
                      <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--lighten-2">keyboard_arrow_up</i></div>
                    </button></span>
                  </div>
                  <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move down</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_DOWN">
                    <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">keyboard_arrow_down</i></div>
                  </button></span>
                </div>
                <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Delete</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-DELETE_ITEM">
                  <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">close</i></div>
                </button></span>
              </div>
              <!---->
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="card-border-light">
    <div class="v-card v-card--flat v-sheet theme--light answer editable closed wrong" data-test="answer">
      <div class="indicator wrong"></div>
      <div class="v-card__text">
        <div class="layout align-top">
          <div class="flex">
            <div class="v-input v-input--selection-controls v-input--radio-group v-input--radio-group--column theme--light">
              <div class="v-input__control">
                <div class="v-input__slot" style="height: auto;">
                  <div role="radiogroup" class="v-input--radio-group__input">
                    <div class="v-radio theme--light">
                      <div class="v-input--selection-controls__input"><input aria-label="Peanut butter" aria-checked="false" role="radio" type="radio" name="v-radio-52" value="1" data-test="answerRadio">
                        <div class="v-input--selection-controls__ripple"></div><i aria-hidden="true" class="v-icon material-icons theme--light">radio_button_unchecked</i>
                      </div><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: relative;">Peanut butter</label>
                    </div>
                  </div>
                </div>
                <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
              </div>
            </div>
            <!---->
            <!---->
          </div>
          <div class="flex">
            <!---->
          </div>
          <div class="spacer"></div>
          <div class="flex">
            <div class="layout justify-end toolbar" data-test="toolbar">
              <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move up</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_UP">
                <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">keyboard_arrow_up</i></div>
              </button></span>
            </div>
            <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move down</span></div><button disabled="disabled" type="button" class="v-btn v-btn--disabled v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_DOWN">
              <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--lighten-2">keyboard_arrow_down</i></div>
            </button></span>
          </div>
          <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Delete</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-DELETE_ITEM">
            <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">close</i></div>
          </button></span>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
</div>
</div> <button type="button" class="mt-3 ml-0 v-btn v-btn--flat theme--light primary--text" data-test="newAnswerBtn">
  <div class="v-btn__content">
    New answer
  </div>
</button></div>
<div class="mt-4">
  <div class="grey--text text--darken-1 mb-3">
    Hints
  </div>
  <!---->
  <div data-test="hint" class="card-border-light">
    <div class="v-card v-card--flat v-sheet theme--light hint closed">
      <div class="v-card__text">
        <div class="layout align-top">
          <div class="flex xs1">
            1
          </div>
          <div class="flex"><span name="fade">
                  It's not healthy
                </span></div>
          <div class="spacer"></div>
          <div class="flex">
            <div class="layout justify-end toolbar">
              <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move up</span></div><button disabled="disabled" type="button" class="v-btn v-btn--disabled v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_UP">
                <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--lighten-2">keyboard_arrow_up</i></div>
              </button></span>
            </div>
            <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move down</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_DOWN">
              <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">keyboard_arrow_down</i></div>
            </button></span>
          </div>
          <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Delete</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-DELETE_ITEM">
            <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">close</i></div>
          </button></span>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
</div>
</div>
<div data-test="hint" class="card-border-light">
  <div class="v-card v-card--flat v-sheet theme--light hint closed">
    <div class="v-card__text">
      <div class="layout align-top">
        <div class="flex xs1">
          2
        </div>
        <div class="flex"><span name="fade">
                  Tasty!
                </span></div>
        <div class="spacer"></div>
        <div class="flex">
          <div class="layout justify-end toolbar">
            <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move up</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_UP">
              <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">keyboard_arrow_up</i></div>
            </button></span>
          </div>
          <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move down</span></div><button disabled="disabled" type="button" class="v-btn v-btn--disabled v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_DOWN">
            <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--lighten-2">keyboard_arrow_down</i></div>
          </button></span>
        </div>
        <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Delete</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-DELETE_ITEM">
          <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">close</i></div>
        </button></span>
      </div>
      <!---->
    </div>
  </div>
</div>
</div>
</div>
</div> <button type="button" class="mt-3 ml-0 v-btn v-btn--flat theme--light primary--text" data-test="newHintBtn">
  <div class="v-btn__content">
    New hint
  </div>
</button></div>
</div>
</div>
</div>
`;
